File: src/main/java/com/ucc/finanzas/service/MetaAhorroService.java
package com.ucc.finanzas.service;

import com.ucc.finanzas.dto.MetaAhorroRequest;
import com.ucc.finanzas.model.MetaAhorro;
import com.ucc.finanzas.model.Usuario;
import com.ucc.finanzas.repository.MetaAhorroRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Permite crear metas de ahorro mensuales y verificar si se cumplieron.
 */
@Service
public class MetaAhorroService {

    private final MetaAhorroRepository metaAhorroRepository;
    private final AuthService authService;
    private final ReporteService reporteService;

    public MetaAhorroService(MetaAhorroRepository metaAhorroRepository,
                             AuthService authService,
                             ReporteService reporteService) {
        this.metaAhorroRepository = metaAhorroRepository;
        this.authService = authService;
        this.reporteService = reporteService;
    }

    public MetaAhorro crearMeta(MetaAhorroRequest request) {
        Usuario usuario = authService.buscarPorId(request.getUsuarioId());

        MetaAhorro meta = new MetaAhorro(
                request.getMontoMeta(),
                request.getMes(),
                request.getAnio(),
                usuario
        );

        if (!meta.validarMonto()) {
            throw new RuntimeException("El monto de la meta debe ser mayor a cero.");
        }

        return metaAhorroRepository.save(meta);
    }

    public boolean verificarCumplimiento(Long usuarioId, int mes, int anio) {
        MetaAhorro meta = metaAhorroRepository
                .findByUsuarioIdAndMesAndAnio(usuarioId, mes, anio)
                .orElseThrow(() -> new RuntimeException(
                        "No hay meta registrada para el mes " + mes + "/" + anio));

        double balance = reporteService.generarReporteMensual(usuarioId, mes, anio).getBalance();
        return meta.verificarCumplimiento(balance);
    }

    public Optional<MetaAhorro> obtenerMetaMes(Long usuarioId, int mes, int anio) {
        return metaAhorroRepository.findByUsuarioIdAndMesAndAnio(usuarioId, mes, anio);
    }

    public List<MetaAhorro> listarPorUsuario(Long usuarioId) {
        return metaAhorroRepository.findByUsuarioId(usuarioId);
    }

    public void eliminarMeta(Long id) {
        if (!metaAhorroRepository.existsById(id)) {
            throw new RuntimeException("Meta no encontrada con id: " + id);
        }
        metaAhorroRepository.deleteById(id);
    }
}

File: frontend/src/app/services/meta-ahorro.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { MetaAhorro, MetaAhorroRequest } from '../models/meta-ahorro.model';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class MetaAhorroService {
  private apiUrl = `${environment.apiUrl}/metas`;

  constructor(private http: HttpClient) {}

  crearMeta(req: MetaAhorroRequest): Observable<MetaAhorro> {
    return this.http.post<MetaAhorro>(this.apiUrl, req);
  }

  obtenerMetaMes(usuarioId: number, mes: number, anio: number): Observable<MetaAhorro> {
    return this.http.get<MetaAhorro>(`${this.apiUrl}/${usuarioId}`, { params: { mes, anio } });
  }

  verificarCumplimiento(usuarioId: number, mes: number, anio: number): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/${usuarioId}/verificar`, { params: { mes, anio } });
  }

  listarPorUsuario(usuarioId: number): Observable<MetaAhorro[]> {
    return this.http.get<MetaAhorro[]>(`${this.apiUrl}/${usuarioId}/todas`);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
